src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/PoolWriter.java

```diff
@@ -250,25 +250,14 @@ void enterInnerClass(ClassSymbol c) {
         }
     }
 
-    /** Enter a value class into the `valueClasses' set.
+    /** Enter a value class into the `preloadClasses' set.
      */
-    void enterValueClass(ClassSymbol c) {
+    void enterPreloadClass(ClassSymbol c) {
         if (c.type.isCompound()) {
             throw new AssertionError("Unexpected intersection type: " + c.type);
         }
         c.complete();
-        if (c.isValueClass() && !c.isPrimitiveClass()) {
```

Comment:
This is fully covered by `requiresPreload` in `Type`. Actually more strict, because of the `if (this.isReferenceProjection()) return true;` condition in the previous method. Is my understanding correct?

Reply:
> This is fully covered by `requiresPreload` in `Type`. Actually more strict, because of the `if (this.isReferenceProjection()) return true;` condition in the previous method. Is my understanding correct?

Yes, you are correct. The check
```
if (this.isReferenceProjection()) return true;
```
covers the requirement for https://bugs.openjdk.java.net/browse/JDK-8280942

The check
```
return this.isValueClass() && !this.isPrimitiveClass();
```
works for "pure" B2 classes. (every B3 is also a B2)